### PR TITLE
feat: adds config file and AuthTokenStorage

### PR DESCRIPTION
**Pull request by fenilli**
This PR extends the current config to a file instead, by a virtual file via `addPluginTemplate` we can import it, and set the options of it with functions for authStorage as #43.

While it does do it, there are considerations and changes I would like to ask from the maintainer @manchenkoff, as one doing it like this will break backwards compatibility, so we could instead have both ways, instead where one is a smaller subset of the other ( module doesn't accept functions, so no auth storage ). And secondly, I would probably use it to change other places, as to not use CSRF when using auth storages, maybe changing the login/logout paths and so on, would like thoughts on it.

### CODE REVIEW

**Comment by fenilli** [comment]
1. Needs to fix the importing the helper function for type safety ( can't be in modules.ts ).

**Comment by jjjrmy** [comment]
a few questions for @manchenkoff ...

Do we want to have the ability to store the CSRF token using the same method?
Basically `useCookie()` would be the default implementation of the AuthStorage.
Then would we have a config variable to select how we are sending the the "token" to the server?
Either as `set-cookie` header or `Authorization` header.
Because also we should not have `credentials: include` when not using Cookie mode.

**Comment by manchenkoff** [comment]
Closes #33

**Comment by manchenkoff** [comment]
Hey @fenilli, thanks a lot for the contribution! I will take a look in a couple of days and give feedback from my side. And also I'll check your questions @jjjrmy 👍

File: src/runtime/httpFactory.ts [hunk 1]
```diff
@@ -77,6 +76,14 @@ export function createHttpClient(): $Fetch {
                 ...options.headers,
             };
 
+            if (config.authTokenStorage) {
```
Comment (manchenkoff) [inline src/runtime/httpFactory.ts:79]: I think it would be a bit overwhelming to have both bearer token and later csrf (in `buildServerHeaders`) since we know that only one strategy should be applied. By the way, I am unsure that it will work for both CSR/SSR, have you checked with some implementation of the storage?

File: src/types.ts [hunk 2]
```diff
@@ -75,9 +75,37 @@ export interface RedirectOptions {
...
+}
+
+/**
+ * Options to be passed to the module plugin.
  */
 export interface SanctumModuleOptions {
+    /**
+     * The path for the configuration file, defaults to "sanctum.config.ts".
+     */
+    configFile?: string;
```
Comment (manchenkoff) [inline src/types.ts:102]: this will break compatibility with the existing configuration, which is not a very preferable way 😄 is there a reason we can't stick to extending the current implementation?

**Comment by manchenkoff** [comment]
> a few questions for @manchenkoff ...
> 
> Do we want to have the ability to store the CSRF token using the same method? Basically `useCookie()` would be the default implementation of the AuthStorage. Then would we have a config variable (mode) to select how we are sending the the "token" to the server? Either as `set-cookie` header or `Authorization` header.

Yes, ideally I would expect different implementations for different scenarios, so we can change the `mode` to one available from enum or something. In the case of Laravel Sanctum, we would have 2 options: `cookie` and `token`, based on the selected mode we should change the whole behavior of the plugin and not conditionally combine them.

**Comment by manchenkoff** [comment]
> or we can have a toggle mode for Cookie vs Auth Token, and use the default callback as localStorage for authToken unless overwritten within the config.

this is very similar to the approach you mentioned, but the problem with `localStorage` is that it does not work with SSR, and the whole point of the library evaporates 😄

**Comment by manchenkoff** [comment]
> This PR extends the current config to a file instead, by a virtual file via `addPluginTemplate` we can import it, and set the options of it with functions for authStorage as #43.

I've already asked about more details on that in the review, but is it the only way we can support callbacks or complex types in the configuration?

> While it does do it, there are considerations and changes I would like to ask from the maintainer @manchenkoff, as one doing it like this will break backwards compatibility, so we could instead have both ways, instead where one is a smaller subset of the other ( module doesn't accept functions, so no auth storage ). 

Sorry, this one is very confusing for me 😄 so which option do you suggest w/o breaking backward compatibility?

> And secondly, I would probably use it to change other places, as to not use CSRF when using auth storages, maybe changing the login/logout paths and so on, would like thoughts on it.

Agree, also mentioned it in one of the previous comments. Since the module configuration has a set of common values, I would assume that you can reuse all things like login/logout endpoints and redirects with both modes - cookie and token. The only difference I see is how CSR/SSR handles the request, so behavior should be changed according to the selected mode in the config. For instance, when we have `mode: cookie` we still request CSRF and pass it as a token, but w/o token checks and the opposite for second `mode: token` - we pass token header w/o trying to enrich request with CSRF-related data.

**Review by manchenkoff — CHANGES_REQUESTED** [review]

**Comment by fenilli** [comment]
> I've already asked about more details on that in the review, but is it the only way we can support callbacks or complex types in the configuration?

Yes sadly it is a limitation of Nuxt config file, because they serialize it before sending to plugins, as stated here [#2822](https://github.com/epicmaxco/vuestic-ui/issues/2822).

> Sorry, this one is very confusing for me 😄 so which option do you suggest w/o breaking backward compatibility?

Yeh I should have made a better explanation there, but basically because we can't pass anything that is not serializable via `nuxt.config` we can't have functions there like what we would need to implement an interface for token saving, so for having backwards compatibility, I can see two scenarios.

1. We would just add the `configFile` property to the module one, and merge both the options from `nuxt.config` and the config file options if passed.
2. We could have a discriminating union, where you can pass to `nuxt.config` the module options, or the `configFile`, for a possible more complete options.

> Agree, also mentioned it in one of the previous comments. Since the module configuration has a set of common values, I would assume that you can reuse all things like login/logout endpoints and redirects with both modes - cookie and token. The only difference I see is how CSR/SSR handles the request, so behavior should be changed according to the selected mode in the config. For instance, when we have `mode: cookie` we still request CSRF and pass it as a header, but w/o token checks and the opposite for second `mode: token` - we pass the token header w/o trying to enrich the request with CSRF-related data.

Agree with this point, not sure about the `mode` tho, one thing we could do is make the storage instead of being a storage implementation, but a HTTP request transformer, you would internally still be able to make storage, but also change the HTTP request when using `get` to pass csrf or auth headers depending on what you're using there, and we would have a default implementation of this using CSRF.

What are your thoughts on these?

File: src/index.d.ts [hunk 3]
```diff
@@ -2,6 +2,6 @@ import { SanctumModuleOptions } from './types';
 
 declare module 'nuxt/schema' {
     interface PublicRuntimeConfig {
-        sanctum: Partial<SanctumModuleOptions>;
+        sanctum: SanctumModuleOptions;
```
[the remark this answers is not in the document]
Comment (fenilli) [reply src/index.d.ts:5]: For sure I will run those and see what warnings they are giving.

File: src/module.ts [hunk 4]
```diff
@@ -18,40 +19,59 @@ export default defineNuxtModule<Partial<SanctumModuleOptions>>({
...
-        },
+        configFile: 'sanctum.config.ts',
     },
 
     setup(options, nuxt) {
         const resolver = createResolver(import.meta.url);
 
-        const publicConfig = nuxt.options.runtimeConfig.public;
-        const userModuleConfig = publicConfig.sanctum;
+        const configBase = resolver.resolve(nuxt.options.rootDir, options.configFile ?? 'sanctum.config');
```
[the remark this answers is not in the document]
Comment (fenilli) [reply src/module.ts:28]: Basically it is meant to resolve paths in the relative path of the users directory project, and not at the module level, that is why we use the `resolve` with the `nuxt.options.rootDir` so we get the users project root directory.

File: src/module.ts [hunk 5]
```diff
@@ -18,40 +19,59 @@ export default defineNuxtModule<Partial<SanctumModuleOptions>>({
...
+
+        addPlugin(resolver.resolve('./runtime/plugin'));
+
+        addPluginTemplate({
```
[the remark this answers is not in the document]
Comment (fenilli) [reply src/module.ts:32]: We can't import things from the users project at build time without having a template be created at runtime that does it for us, this template creates a plugin that has that config file imported.

File: src/module.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (manchenkoff) [reply src/module.ts:28]: I see, different names of config file are a bit confusing for me, I would suggest using one static name, the same as `tailwind.config.ts` that is fixed for the module w/o the ability to change it. It provides less flexibility but will keep usage simpler for the user. What do you think?

File: src/runtime/composables/useSanctumConfig.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import type { SanctumConfigOptions } from '../../types';
+import { useNuxtApp } from '#app';
+
+export const useSanctumConfig = (): SanctumConfigOptions => {
```
[the remark this answers is not in the document]
Comment (fenilli) [reply src/runtime/composables/useSanctumConfig.ts:4]: You do have a point there, you could say that the user could change the values from the config, so do we if we created as state, and a user could add this to their plugins, but I guess it is more inline with other packages to use config files for these kinda of things.

File: src/types.ts (hunk 2 above)
Comment (fenilli) [reply src/types.ts:102]: We can, it would be just a bit more work with merging but totally possible.

File: src/module.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (fenilli) [reply src/module.ts:28]: For sure that might be easier, in that case we can drop the change in the options interface, and just get from root, this would make it easier for backwards compatibility too.

**Comment by manchenkoff** [comment]
> Yes sadly it is a limitation of Nuxt config file, because they serialize it before sending to plugins, as stated here [#2822](https://github.com/epicmaxco/vuestic-ui/issues/2822).
> 

Got it, looks good to me!

> Yeh I should have made a better explanation there, but basically because we can't pass anything that is not serializable via `nuxt.config` we can't have functions there like what we would need to implement an interface for token saving, so for having backwards compatibility, I can see two scenarios.
> 
> 1. We would just add the `configFile` property to the module one, and merge both the options from `nuxt.config` and the config file options if passed.
> 2. We could have a discriminating union, where you can pass to `nuxt.config` the module options, or the `configFile`, for a possible more complete options.

I think I see the idea, so for me, as a user of this package I would say that the crucial part is the simplicity of the configuration and easy integration - what if we add `tokenStorage` only as a separate config file? In this case, everybody who uses the existing module with cookie-based authentication won't need to change anything, but for those interested in the token storage it will be enough to create a new file and define the implementations. This will allow us to merge the functionality w/o breaking compatibility and we'll be able to refactor/simplify it later.

> Agree with this point, not sure about the `mode` tho, one thing we could do is make the storage instead of being a storage implementation, but a HTTP request transformer, you would internally still be able to make storage, but also change the HTTP request when using `get` to pass csrf or auth headers depending on what you're using there, and we would have a default implementation of this using CSRF.

Yeah, it is not necessary to have `mode` there, it was just an example for the concept. This way with transformers also sounds good.

To sum up, my main point is to keep this module as simple as possible for the end user and keep backward compatibility as much as possible for now

File: src/types.ts (hunk 2 above)
Comment (manchenkoff) [reply src/types.ts:102]: answered, nuxt limitation

File: src/runtime/httpFactory.ts (hunk 1 above)
Comment (manchenkoff) [reply src/runtime/httpFactory.ts:79]: now I'm curious if it's even possible to use token storage with SSR? 😄 I've never used it that way

File: src/module.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (manchenkoff) [reply src/module.ts:32]: clear, thanks!

File: src/runtime/composables/useSanctumConfig.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (manchenkoff) [reply src/runtime/composables/useSanctumConfig.ts:4]: makes sense, probably omitting this composable from the documentation would be enough 😄

File: src/index.d.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (fenilli) [reply src/index.d.ts:5]: I fixed this, sadly I have to be careful " using windows, all files are modified by the prettier ".

File: src/runtime/httpFactory.ts (hunk 1 above)
Comment (jjjrmy) [reply src/runtime/httpFactory.ts:79]: I’ll have to look into that. but I think with a mobile app there is no SSR? I haven’t been able to test this with SSR because of previous issues. But I will research this, I think there are some things like asyncData for sharing data between Serve and Client

Comment (manchenkoff) [reply src/runtime/httpFactory.ts:79]: I'd appreciate that, thanks!

File: playground/sanctum.config.ts [hunk 7]
```diff
@@ -0,0 +1,4 @@
+export default {
+    baseUrl: 'http://localhost:80',
+    userStateKey: 'sanctum.user.identity',
```
[the remark this answers is not in the document]
Comment (fenilli) [reply playground/sanctum.config.ts]: Removed that from the config, just cause it.

File: src/runtime/plugin.ts [hunk 8]
```diff
@@ -18,21 +18,16 @@ function handleIdentityLoadError(error: Error) {
...
             handleIdentityLoadError(error as Error);
         }
     }
 
-    return {
-        provide: {
-            sanctumClient: client,
-        },
-    };
+    nuxtApp.provide('sanctumClient', client);
```
Comment (manchenkoff) [inline src/runtime/plugin.ts:32]: is there any difference from the functional perspective?

**Review by manchenkoff — APPROVED** [review]

**Comment by fenilli** [comment]
> I think I see the idea, so for me, as a user of this package I would say that the crucial part is the simplicity of the configuration and easy integration - what if we add `tokenStorage` only as a separate config file? In this case, everybody who uses the existing module with cookie-based authentication won't need to change anything, but for those interested in the token storage it will be enough to create a new file and define the implementations. This will allow us to merge the functionality w/o breaking compatibility and we'll be able to refactor/simplify it later.

I added back the options runtimeConfig and sanctum to the module, and made defu merge later with the config file. so now it should be backwards compatible with the addition of config files for complex configuration.

> Yeah, it is not necessary to have `mode` there, it was just an example for the concept. This way with transformers also sounds good.
> 
> To sum up, my main point is to keep this module as simple as possible for the end user and keep backward compatibility as much as possible for now

Still need to think a bit on what should we do now to make it simpler for those "modes" to happen based on what, I think a transformer might be good, just need to think how would it transform and where it should do it, both for request and response, as we might need information from the response to be passed somewhere, so it can be saved ( for the token case ).

I was probably thinking in changing the AuthStorage idea to something more in line with changing the request/response headers. as the storage in itself is not necessary, it can be internal to the users project, he would just need to get the response from login to do anything he wants, and change the request headers, to send anything he needs for its implementation.

And we would use the cookies csrf as base if nothing is passed to it.

File: src/runtime/composables/useSanctumConfig.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (fenilli) [reply src/runtime/composables/useSanctumConfig.ts]: A user could use it too, not sure why, but I added to make sure a Readonly type as return to it.

File: src/runtime/plugin.ts (hunk 8 above)
Comment (fenilli) [reply src/runtime/plugin.ts:32]: To be honest I'm not sure if there are differences, I had some problems with the return when inside the template ( it might been another problem that I fixed and didn't make any difference that syntax ), so I changed so both use the same syntax, I can change it back and see if both work as functional instead.

**Comment by fenilli** [comment]
Also do you know how to export other functions from a Nuxt module? we can't really export from the main entry point, nuxt doesn't like that, and a helper function like `defineSanctumConfig` would be very helpful for type safety in the config file.

**Comment by manchenkoff** [comment]
> Also do you know how to export other functions from a Nuxt module? we can't really export from the main entry point, nuxt doesn't like that, and a helper function like `defineSanctumConfig` would be very helpful for type safety in the config file.

I don't but I looked it up in the repo you shared previously and it seems that they exported the utility function in the [main file here](https://github.com/epicmaxco/vuestic-ui/blob/c1084e9451d6a21d68403952c1eacef65708d644/packages/ui/src/main.ts#L26). If it doesn't work, I would expect just something like `Partial<SanctumConfig>` in the config file

**Comment by fenilli** [comment]
> I don't but I looked it up in the repo you shared previously and it seems that they exported the utility function in the [main file here](https://github.com/epicmaxco/vuestic-ui/blob/c1084e9451d6a21d68403952c1eacef65708d644/packages/ui/src/main.ts#L26). If it doesn't work, I would expect just something like `Partial<SanctumConfig>` in the config file

they export at the main of the ui package, not the modules nuxt one, that is why it works, for some reason nuxt doesn't like when you export anything else in their module main export.

File: src/runtime/plugin.ts (hunk 8 above)
Comment (manchenkoff) [reply src/runtime/plugin.ts:32]: The only concern I have that it might be different order of execution during plugin registration since we use `nuxtApp` directly instead of returning plugin to be registered via `provide`, it would be great if you can double-check that it works with the previous version as well

**Comment by fenilli** [comment]
Ok after thinking about a lot, there is one thing I might like and want opnions about @jjjrmy and @manchenkoff.

As of now this package is meant for basically plug and play default behaviour, and its really good for that, with that in mind, I like the simple JSON options that it has, but its meant only for cookies (csrf), but laravel sanctum also accepts token based calls, so it might be a good thing to accept also by default a simple token based approach.

What I was thinking on how to do it, as simple as possible and to keep backwards compatible, was to do a discriminated union for configurations, and let the user choose in the config via a manager if they want from cookie vs token, it will also enable to keep the config exactly the same but with one additional thing that users may tap into it, the storage and transformer.

For cookies is simple, just add storage as config for only the config file, use the base `cookie`, and if someone wants he can instead pass a function that follows the storage interface ( add, remove, get ) all the configuration would stay the same outside of those two new properties.

For tokens, we would make changes to the current code to use each manager correctly, and each manager would do its job to add/remove things from the headers, and token would have two main default strategies in the storage, `local` and `cookies`, but it can also pass a storage itself for mobile for example.

The transfomer would be useful for users to at the end of requests change the request options if needed for any reason they might think, making this package really simple if needed, but also accept complex instructions if needed.

**Comment by manchenkoff** [comment]
Hey @fenilli 

> What I was thinking on how to do it, as simple as possible and to keep backwards compatible, was to do a discriminated union for configurations, and let the user choose in the config via a manager if they want from cookie vs token, it will also enable to keep the config exactly the same but with one additional thing that users may tap into it, the storage and transformer.

Your suggestion sounds good, I don't mind having it like that. The only concern I have is the compatibility with the current `fetch` implementation since there is no way to modify `onRequest` and `onResponse` interceptors after instantiation, please keep that in mind when designing custom transformers functionality.

Besides that, we are good to go! Feel free to push changes when you have time to discuss further! And thanks a lot for your participation, I appreciate that! 😎

**Comment by manchenkoff** [comment]
Closing for now, we can re-open this one once we have updates or a new one

**Comment by fenilli** [comment]
Sorry fo the delay, just having a lot of things to do, but I will try to make a new one this week, also thinking if we should have a way to have one or another, or possibility to have both at the same time.
